### PR TITLE
[PO-77] Use mock SC for CA Voting to create and validate blocks

### PR DESCRIPTION
**Pull request by javierdiaz72**
The purpose of this PR is to use the mock smart contract for CA voting in order to check whether the node is the slot leader for the current slot and to validate blocks.

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/nodebuilder/NodeBuilder.scala [hunk 1]
```diff
@@ -86,9 +86,9 @@ trait OuroborosConfigBuilder {
   lazy val ouroborosConfig = OuroborosConfig(Config.config)
 }
 
-trait ElectionManagerBuilder {
-  self: OuroborosConfigBuilder =>
-  lazy val electionManager = ElectionManagerImpl(ouroborosConfig)
+trait CertificateAuthorityManagerBuilder {
+  self: LedgerBuilder with BlockchainBuilder with OuroborosConfigBuilder with BlockchainConfigBuilder =>
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/nodebuilder/NodeBuilder.scala]: Minor detail: Is `LedgerBuilder` needed here?

File: src/main/scala/io/iohk/ethereum/vm/utils/Contract.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    }
+  }
+
+  private def callWithSignature(signature: String)(args: Any*): ContractMethodCall[W, S] = {
+    val signatureHash = ByteString(crypto.kec256(signature.getBytes)).take(4)
+    val callData = args.map(parseArg).foldLeft(signatureHash)(_ ++ _)
+    ContractMethodCall(this, callData, blockHeader, evmConfig)
+  }
+
+  private def parseArg(arg: Any): ByteString = arg match {
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/vm/utils/Contract.scala]: Minor detail: Maybe we should make this function public in a companion object as it's a copy from the same function in `EvmTestEnv.scala`?

File: src/evmTest/scala/io/iohk/ethereum/vm/utils/EvmTestEnv.scala [hunk 3]
```diff
@@ -83,55 +80,36 @@ trait EvmTestEnv {
...
+    def callMethod(methodName: String)(args: Any*): MockContractMethodCall = {
+      val methodCall = contract.callMethod(methodName)(args: _*)
+      new MockContractMethodCall(this, methodCall)
     }
 
     def storage: MockStorage = world.storages(address)
   }
 
-  class ContractMethodCall(contract: Contract, callData: ByteString) {
+  class MockContractMethodCall(mockContrack: MockContract,
```
Comment (ntallar) [inline src/evmTest/scala/io/iohk/ethereum/vm/utils/EvmTestEnv.scala]: Minor detail: Maybe we should remove the `mockContract` parameter as it's not being used?

**Review by ntallar — COMMENTED** [review]
Apart from this minor details, LGTM!

File: src/main/scala/io/iohk/ethereum/nodebuilder/NodeBuilder.scala (hunk 1 above)
Comment (mirkoAlic) [reply src/main/scala/io/iohk/ethereum/nodebuilder/NodeBuilder.scala]: :+1:

File: src/evmTest/scala/io/iohk/ethereum/vm/utils/EvmTestEnv.scala (hunk 3 above)
Comment (mirkoAlic) [reply src/evmTest/scala/io/iohk/ethereum/vm/utils/EvmTestEnv.scala]: sure!

File: src/main/scala/io/iohk/ethereum/vm/utils/Contract.scala (hunk 2 above)
Comment (mirkoAlic) [reply src/main/scala/io/iohk/ethereum/vm/utils/Contract.scala]: :100:

**Review by ntallar — APPROVED** [review]
LGTM!

**Pull request merged**